repair/repair.cc

```diff
@@ -2116,7 +2116,6 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
         auto myhostid = erm->get_token_metadata_ptr()->get_my_id();
         auto myip = erm->get_topology().my_address();
         auto mydc = erm->get_topology().get_datacenter();
```

Comment:
> since there's no global primary replica with tablets

why not?

If we run `nodetool repair` on all nodes, won't this over-repair? For a given tablets, there will be nr_datacenters primary replicas.

Reply:
> > since there's no global primary replica with tablets
> 
> why not?

vnode allocation has clear primary owner in token metadata.
tablet allocation just allocates arbitrarily in each dc, based on the network topology rf config, and we only shuffle nodes within racks and rotate racks within DCs so to balance the first replica in each dc per tablet, while keeping rack pairing order for MV.

We can rorate also DC order for each tablet to facilitate global primary replica, or choose a primary DC by some fair algorithm in repair.

fully-automated tablet repair will just solve this without relying on -pr option.

> 
> If we run `nodetool repair` on all nodes, won't this over-repair? For a given tablets, there will be nr_datacenters primary replicas.

yes.  one could provide explicit token ranges based on the tablet map to overcome this.

Reply:
> > > since there's no global primary replica with tablets
> > 
> > 
> > why not?
> 
> vnode allocation has clear primary owner in token metadata. tablet allocation just allocates arbitrarily in each dc, based on the network topology rf config, and we only shuffle nodes within racks and rotate racks within DCs so to balance the first replica in each dc per tablet, while keeping rack pairing order for MV.
> 
> We can rorate also DC order for each tablet to facilitate global primary replica, or choose a primary DC by some fair algorithm in repair.

Then we should.


> 
> fully-automated tablet repair will just solve this without relying on -pr option.


Sure, but we want something simple first.

> 
> > If we run `nodetool repair` on all nodes, won't this over-repair? For a given tablets, there will be nr_datacenters primary replicas.
> 
> yes. one could provide explicit token ranges based on the tablet map to overcome this.

I don't understand. I want `nodetool repair -pr` to work.

Reply:
> since there's no global primary replica with tablet

Why not? No matter how we allocate replica nodes for a given tablet token range. We can always choose one of the node as primary replica. So there is always a vald primary replica, no matter what.

Currently, we choose the first node in the replica set as the primary replica. We can argue this will cause primary replicas not being distributed evenly among the DCs, but this is not going to change the fact, there is always a primary replica for a given tablet.